### PR TITLE
Feat/track metrics

### LINKED ISSUE #768: Track metrics in the bounties platform (Postgres)
I'd like for us to start tracking key events. This will require more scope but here are a few to start:

Feature brief: Bounties are new way of working, so we want to learn as much as we can about the process of creating and completing bounties. Metrics provide a top down factual view of results. User interviews would provide more qualitative information.

Architecture: Redis vs. gocache?

### Metrics v1
- [x] When a bounty is posted
https://github.com/stakwork/sphinx-tribes/blob/master/db/structs.go#L345
- [x]  When someone is assigned a bounty
check if were using this currently?
https://github.com/stakwork/sphinx-tribes/blob/master/db/structs.go#L360
- [x] When someone is paid
_I think a new field may need to be added, we might want to distinguish between paid through an org/the platform or just marked paid_
- [x] When an org is created
https://github.com/stakwork/sphinx-tribes/blob/master/db/structs.go#L406C2-L406C2
- [x] Amount in a bounty
- [x] Total amount paid over "x" period of time
- [x] Total bounties paid over "x" period of time
- [x] Total bounties assigned over "x" period of time
- [x] Total bounties posted over "x" period of time
- [x] % of bounties completed over "x" period of time
- [x] % of bounties paid over over "x" period of time

### v.2 for super admin
- [ ] Time to value for bounty provider: From account creation to pay out.
   _make query to lookup from bounties table backwards for their owner_id=pubkey and paid=true and then take the paid date - profile creation date_
- [ ] Time to value for bounty hunter: From account creation to getting paid.
   _make query to lookup from bounties table backwards for their assigne_pubkey=pubkey and paid=true and then take the paid date - profile creation date_

### PR DESCRIPTION
**Pull request by elraphty**
## Describe your changes

This PR adds the metrics functionality

## Issue ticket number and link

Closes #768

## Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Checklist before requesting a review
- [ ] I have performed a self-review of my code
- [ ] I have tested on Chrome and Firefox
- [ ] I have tested on a mobile device

### CODE REVIEW

File: db/redis.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+func InitRedis() {
+	redisURL := os.Getenv("REDIS_URL")
+
+	if redisURL == "" {
+		dbInt, _ := utils.ConvertStringToInt(os.Getenv("REDIS_DB"))
+		RedisClient = redis.NewClient(&redis.Options{
+			Addr:     os.Getenv("REDIS_HOST"),
+			Username: os.Getenv("REDIS_USER"),
+			Password: os.Getenv("REDIS_PASS"),
```
Comment (kevkevinpal) [inline db/redis.go:24]: these new env vars we might want to add to the README.md or somewhere similar

File: routes/metrics.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+func MetricsRoutes() chi.Router {
+	r := chi.NewRouter()
+
+	r.Group(func(r chi.Router) {
+		r.Use(auth.PubKeyContext)
+
+		r.Post("/payment_metrics", handlers.PaymentMetrics)
+		r.Post("/people_created", handlers.PeopleMetrics)
+		r.Post("/organization_created", handlers.OrganizationtMetrics)
+		r.Post("/bounty_metrics", handlers.BountyMetrics)
```
Comment (kevkevinpal) [inline routes/metrics.go:18]: we can probably rename these to
`/metrics/paymet`
`/metrics/people`
`/metrics/organization`
`/metrics/bounty`

adding metric and created seems a bit redundant to me let me know if you have differing thoughts

File: routes/metrics.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/go-chi/chi"
+	"github.com/stakwork/sphinx-tribes/auth"
+	"github.com/stakwork/sphinx-tribes/handlers"
+)
+
+func MetricsRoutes() chi.Router {
+	r := chi.NewRouter()
+
+	r.Group(func(r chi.Router) {
+		r.Use(auth.PubKeyContext)
```
Comment (kevkevinpal) [inline routes/metrics.go:13]: I'm assuming we'll add the admin middleware in the future?

File: utils/helpers_test.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetDateDaysDifference(t *testing.T) {
+	created := 1700238956
+
+	testDate := time.Now()
+	days := GetDateDaysDifference(int64(created), &testDate)
+	assert.NotEqual(t, 0, days)
```
Comment (kevkevinpal) [inline utils/helpers_test.go:15]: can we add an assertion for `Equal` and try and do some edge cases aswell like if the date is negative or malformed

Comment (kevkevinpal) [reply utils/helpers_test.go:15]: otherwise its good to see some tests haha

File: handlers/bounties.go [hunk 5]
```diff
@@ -137,9 +137,9 @@ func MigrateBounties(w http.ResponseWriter, r *http.Request) {
 				migrateBountyFinal.Award = Award
 			}
 
-			Price, ok5 := migrateBounty["price"].(string)
+			Price, ok5 := migrateBounty["price"].(uint)
 			if !ok5 {
-				migrateBountyFinal.Price = "0"
+				migrateBountyFinal.Price = 0
```
Comment (kevkevinpal) [inline handlers/bounties.go:142]: so I think this only happened when we did that initial migration so I don't think this will modify any existing bounties so they will continue to be strings. Not sure if we should do this change, we should honestly remove this migration code aswell since if we run it again we will have duplicate bounties which is not good

File: frontend/app/src/people/main/FocusView.tsx [hunk 6]
```diff
@@ -186,6 +186,10 @@ function FocusedView(props: FocusViewProps) {
           newBody.description = description;
         }
 
+        if (newBody.price) {
+          newBody.price = Number(newBody.price);
+        }
```
Comment (kevkevinpal) [inline frontend/app/src/people/main/FocusView.tsx:191]: This fixes the issue of the bounty price being a string right? not sure if we want to return as a number from the backend so the frontend doesn't have to do this. But that may be a larger change

File: db/structs.go [hunk 7]
```diff
@@ -355,32 +355,37 @@ type Client struct {
...
+	EstimatedSessionLength   string         `json:"estimated_session_length"`
+	EstimatedCompletionDate  string         `json:"estimated_completion_date"`
+	Created                  int64          `json:"created"`
+	Updated                  *time.Time     `json:"updated"`
+	AssignedDate             *time.Time     `json:"assigned_date,omitempty"`
+	CompletionDate           *time.Time     `json:"completion_date,omitempty"`
+	MarkAsPaidDate           *time.Time     `json:"mark_as_paid_date,omitempty"`
+	PaidDate                 *time.Time     `json:"paid_date,omitempty"`
+	PaidDateDifference       int64          `gorm:"type:bigint;not null default:'0'" json:"paid_date_difference,omitempty"`
+	CompletionDateDifference int64          `gorm:"type:bigint;not null default:'0'" json:"completion_date_difference,omitempty"`
```
Comment (kevkevinpal) [inline db/structs.go:387]: are we saving this in the db not sure if we should do that since this can simply be computed by two vars in here

File: db/structs.go [hunk 8]
```diff
@@ -355,32 +355,37 @@ type Client struct {
...
+	OneSentenceSummary       string         `json:"one_sentence_summary"`
+	EstimatedSessionLength   string         `json:"estimated_session_length"`
+	EstimatedCompletionDate  string         `json:"estimated_completion_date"`
+	Created                  int64          `json:"created"`
+	Updated                  *time.Time     `json:"updated"`
+	AssignedDate             *time.Time     `json:"assigned_date,omitempty"`
+	CompletionDate           *time.Time     `json:"completion_date,omitempty"`
+	MarkAsPaidDate           *time.Time     `json:"mark_as_paid_date,omitempty"`
+	PaidDate                 *time.Time     `json:"paid_date,omitempty"`
+	PaidDateDifference       int64          `gorm:"type:bigint;not null default:'0'" json:"paid_date_difference,omitempty"`
```
Comment (kevkevinpal) [inline db/structs.go:386]: same with this not sure if it is needed since we can simply compute this value when needed, it is just a simple subtraction

File: db/structs.go [hunk 9]
```diff
@@ -631,6 +642,17 @@ type Meme struct {
...
 
+type BountyMetrics struct {
+	BountiesPosted         int64 `json:"bounties_posted"`
+	BountiesPaid           int64 `json:"bounties_paid"`
+	BountiesPaidPercentage uint  `json:"bounties_paid_average"`
+	SatsPosted             uint  `json:"sats_posted"`
+	SatsPaid               uint  `json:"sats_paid"`
+	SatsPaidPercentage     uint  `json:"sats_paid_percentage"`
+	AveragePaid            uint  `json:"average_paid"`
+	AverageCompleted       uint  `json:"average_completed"`
```
Comment (kevkevinpal) [inline db/structs.go:653]: is there a reason why two of them are `int64` and the rest are `uint`

File: db/redis.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+		if err != nil {
+			fmt.Println("REDIS URL CONNECTION ERROR ===", err)
+		}
+
+		RedisClient = redis.NewClient(opt)
+	}
+}
+
+func SetValue(key string, value interface{}) {
+	err := RedisClient.Set(ctx, key, value, 6*time.Hour).Err()
```
Comment (kevkevinpal) [inline db/redis.go:38]: can we make `6*time.Hour` a var so it is more obvious incase we want to change that value in the future?

File: db/redis.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+		err := RedisClient.HSet(ctx, key, k, v).Err()
+		if err != nil {
+			fmt.Println("REDIS SET MAP ERROR :", err)
+		}
+	}
+	RedisClient.Expire(ctx, key, 6*time.Hour)
+}
+
+func GetMap(key string) map[string]string {
+	values := RedisClient.HGetAll(ctx, key).Val()
```
Comment (kevkevinpal) [inline db/redis.go:64]: is it not possible to get an error here?

File: db/metrics.go [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+import (
+	"fmt"
+	"math"
+	"net/http"
+
+	"github.com/stakwork/sphinx-tribes/utils"
+)
+
+func (db database) TotalPeopleByDateRange(r PaymentDateRange) int64 {
```
Comment (kevkevinpal) [inline db/metrics.go:11]: I dont see this being used anywhere

File: db/metrics.go [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+	paidCount := DB.PaidDifferenceCount(r)
+	if paidCount != 0 && paidSum != 0 {
+		avg := paidSum / uint(paidCount)
+		avgDays := math.Round(float64(avg))
+		return uint(avgDays)
+	}
+	return 0
+}
+
+func (db database) CompletedDifferenceSum(r PaymentDateRange) uint {
```
Comment (kevkevinpal) [inline db/metrics.go:100]: I dont see this being used anywhere

File: db/metrics.go [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+	var sum uint
+	db.db.Model(&Bounty{}).Where("completion_date_difference != ?",
+		"").Where("created >= ?", r.StartDate).Where("created <= ?", r.EndDate).Select("SUM(completion_date_difference)").Row().Scan(&sum)
+	return sum
+}
+
+func (db database) CompletesDifferenceCount(r PaymentDateRange) int64 {
```
Comment (kevkevinpal) [inline db/metrics.go:107]: I dont see this being used anywhere

File: routes/metrics.go (hunk 3 above)
Comment (elraphty) [reply routes/metrics.go:13]: Yes, we will.

File: config/config.go [hunk 15]
```diff
@@ -29,6 +30,9 @@ func InitConfig() {
 	RelayUrl = os.Getenv("RELAY_URL")
 	MemeUrl = os.Getenv("MEME_URL")
 	RelayAuthKey = os.Getenv("RELAY_AUTH_KEY")
+	AdminStrings := os.Getenv("SUPER_ADMINS")
```
[the remark this answers is not in the document]
Comment (kevkevinpal) [reply config/config.go:33]: @elraphty can you change this value to just `ADMINS` aswell `SUPER_ADMINS` seems a bit much especially since we don't even have the concept of admins implemented yet

**Comment by ecurrencyhodler** [comment]
Great job Raph. This was a big one that unblocks a lot of work. 

@elraphty did we open up endpoints? If so which ones? 

Here is the issue related to that as well: https://github.com/stakwork/sphinx-tribes/issues/1007
